hivemind/dht/node.py

```diff
@@ -138,12 +149,28 @@ async def create(
         self.cache_refresh_evt = asyncio.Event()
         self.cache_refresh_task = None
 
-        self.protocol = await DHTProtocol.create(self.node_id, bucket_size, depth_modulo, num_replicas, wait_timeout,
-                                                 parallel_rpc, cache_size, listen, listen_on, endpoint,
-                                                 record_validator, **kwargs)
-        self.port = self.protocol.port
+        if p2p is None:
+            if not kwargs.get('use_ipfs'):
+                kwargs['initial_peers'] = initial_peers
+            p2p = await P2P.create(**kwargs)
+            self._should_shutdown_p2p = True
+        else:
+            if kwargs:
+                raise ValueError(
+                    f'**kwargs in DHTNode.create() should be empty if hivemind.p2p.P2P instance is provided'
+                    f'in the constructor. Got kwargs = {kwargs} instead. '
+                    f'You may have a typo in a DHTNode.create() parameter name')
+            self._should_shutdown_p2p = False
+        self.p2p = p2p
+
+        self.protocol = await DHTProtocol.create(
+            p2p, self.node_id, bucket_size, depth_modulo, num_replicas, wait_timeout,
+            parallel_rpc, cache_size, listen, record_validator, authorizer)
+        self.endpoint = p2p.id
 
         if initial_peers:
+            initial_peers = {Endpoint.from_base58(Multiaddr(item)['p2p']) for item in initial_peers}
```

Comment:
The set is used to remove duplicates, since it is likely that we have multiple maddrs for each initial peer (however, at this stage, all maddrs are already added to libp2p routing tables, and we only need to iterate over different PeerIDs).